src/Connector/CachingConnector.php

```diff
@@ -26,23 +28,48 @@ public function __construct(CacheItemPoolInterface $cache = null)
         $this->cache = $cache ?: new MemoryCache;
     }
 
-    public function fetch($source, EncapsulatedOptions $options = null)
-    {
+    /**
+     * @param string $source
+     * @param EncapsulatedOptions|null $options
+     * @param CacheKeyGeneratorInterface $cacheKeyGenerator
+     * @return mixed
+     * @throws InvalidCacheKeyException
+     */
+    public function fetch(
+        $source,
+        EncapsulatedOptions $options = null,
+        CacheKeyGeneratorInterface $cacheKeyGenerator = null
```

Comment:
I believe this is fundamentally an incorrect approach. Method parameters should be used for runtime data concerns but what we are dealing with here is a configuration time concern. That is, runtime data is data we expect to change for every method invocation whereas configuration data is typically set once per script lifetime. Therefore I recommend re-factoring the cache key generator is a getter/setter pair.